common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java

```diff
@@ -95,19 +110,140 @@ private ITestRun getTestRun() {
+    }
+
+    @Override
+    public void testSuccess(String testName) {
+        inspectWithLifeCycle(InspectionStrategy.WhenType.TEST_SUCCESS, testName);
+    }
+
+    @Override
+    public void testFailure(String testName) {
+        inspectWithLifeCycle(InspectionStrategy.WhenType.TEST_FAILURE, testName);
+    }
+
+    public void testTearDown(DeviceInfo deviceInfo) {
+        ITestRun testRun = getTestRun();
+
+        List<PerformanceTestResult> resultList = parseForTestRun(testRun);
+        savePerfTestResults(resultList);
+
+        List<ScheduledFuture<?>> timerList = inspectPerformanceTimerMap.get(testRun.getId());
+        if (timerList != null) {
+            for (ScheduledFuture<?> timer : timerList) {
+                timer.cancel(true);
+            }
+        }
+
+        inspectPerformanceTimerMap.remove(testRun.getId());
+        testLifeCycleStrategyMap.remove(testRun.getId());
+        testRunPerfResultMap.remove(testRun.getId());
+
+        //TODO Android battery: adb shell dumpsys battery reset using Device info
+    }
+
+    private void inspectWithLifeCycle(InspectionStrategy.WhenType whenType, String description) {
+        List<InspectionStrategy> strategyList = testLifeCycleStrategyMap.get(getTestRun().getId());
+        if (strategyList == null) return;
+
+        for (InspectionStrategy inspectionStrategy : strategyList) {
+            if (inspectionStrategy == null) continue;
+            PerformanceInspection inspection = inspectionStrategy.inspection;
+            if (inspectionStrategy.when == null) continue;
+
+            if (whenType == InspectionStrategy.WhenType.TEST_RUN_STARTED) {
+                // initialize inspector
+                PerformanceInspection initialInspection = new PerformanceInspection(
+                        InspectionStrategy.WhenType.TEST_RUN_STARTED.name(),
+                        inspection.inspectorType, inspection.appId, inspection.deviceIdentifier, true);
+                inspect(initialInspection);
+            } else if (inspectionStrategy.when.contains(whenType)) {
+                PerformanceInspection lifeCycleInspection = new PerformanceInspection(
+                        whenType.name() + "-" + description,
+                        inspection.inspectorType, inspection.appId, inspection.deviceIdentifier, false);
+                inspect(lifeCycleInspection);
+            }
+        }
+    }
+
+    /**
+     * For giving inspection return the inspection with device id that related to test run
+     */
+    private PerformanceInspection getDevicePerformanceInspection(PerformanceInspection inspection) {
+        if (getTestRun() instanceof TestRun) {
+            TestRun testRun = (TestRun) getTestRun();
+            return new PerformanceInspection(inspection.description, inspection.inspectorType, inspection.appId,
+                    // For windows inspector, the deviceIdentifier is useless
+                    testRun.getDeviceSerialNumber(), inspection.isReset);
```

Comment:
How about exposing getDeviceSerialNumber in ITestRun?

Reply:
Great idea. Updated